clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2773,8 +2773,7 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     TypeSourceInfo *TSInfo =
         KernelObj->isLambda() ? KernelObj->getLambdaTypeInfo() : nullptr;
     auto Type = QualType(KernelObj->getTypeForDecl(), 0);
-    if (KernelObj->isLambda())
-      Type->getAsRecordDecl()->setAnonymousStructOrUnion(true);
+    Type->getAsRecordDecl()->setAnonymousStructOrUnion(KernelObj->isLambda());
```

Comment:
Not 'isLambda'.  we should only be making this type an anon-struct-or-union if the Kernelobj is.  So the condition should be: `KernelObj->isAnonymousStructOrUnion()`.

If this isn't the case, this is a misguided attempt at whatever it is doing.

Reply:
This is not the case, so I need a new guidance here on variable declaration. 

SYCL device compiler is supposed to convert this code:
```c++
inline namespace cl {
namespace sycl {
namespace detail {
template <typename KernelName, typename KernelType>
ATTR_SYCL_KERNEL void kernel_parallel_for_work_group(const KernelType &KernelFunc, kernel_handler kh) {
  KernelFunc(group<1>(), kh);
}
} // namespace detail
} // namespace sycl
} // namespace cl
```
to something like this:
```c++
__kernel void <compiler-generated-mangled-kernel-name>(<kernel parameters>) {
    static <function object type> <name?>; // note: lambda closure type doesn't have a default constructor(!)
    <name>.__init(<kernel parameters>); // object fields initialization
    <name>(); // call operator executes kernel body
    <name>.__finalize(); // finalization is required for stream object
}
```

If understand it correctly the code here declares a variable of type deduced from the template parameter `KernelType`. Is this a valid declaration if `KernelType` is a closure type? I can't write an equivalent code in C++, which can be semantically correct.
What should be changed in this declaration, so that variable type can be mangled? I'm not sure what is wrong right now, but my current hypothesis - IndentifierInfo is invalid (i.e. `KernelObj->getIdentifier()` we pass to `VarDecl::Create`).

Reply:
I don't think the latter is possible: Lambdas are not really constructible in that method and should be constructed in a single-line.  So the __init/__finalize is not really possible here (since they aren't going to be legal functions of a lambda).  We could perhaps place the sub-calls to __init and __finalize in a generated lambda-constructor, however I'm not sure how plausible that is in a SEMA implementation of SYCL.  

The type of the object type SHOULD be mangle-able always, as it is a type.  I could definitely see the VarDecl being created without an identifier being the problem (perhaps debug by giving it a made-up name?), but I would imagine the type itself would be mangleable despite the variable itself not having a name.